src/crypto/spki.rs

```diff
@@ -46,44 +44,50 @@ pub trait SubjectPublicKeyInfoExt {
 
 impl<'a> SubjectPublicKeyInfoExt for SubjectPublicKeyInfo<'a> {
     fn verify(&self, body: &[u8], algo: AlgorithmIdentifier<'_>, sign: &[u8]) -> Result<()> {
-        let alg: &'static dyn VerAlg = match (self.algorithm.oids()?, (algo.oid, algo.parameters)) {
-            ((ECPK, Some(P256)), ES256) => &ECDSA_P256_SHA256_ASN1,
-            ((ECPK, Some(P384)), ES384) => &ECDSA_P384_SHA384_ASN1,
+        match (self.algorithm.oids()?, (algo.oid, algo.parameters)) {
+            ((ECPK, Some(P256)), ES256) => {
+                use p256::ecdsa::signature::Verifier;
+                let vkey = p256::ecdsa::VerifyingKey::from_sec1_bytes(self.subject_public_key)?;
+                let sig = p256::ecdsa::Signature::from_der(sign)?;
+                return Ok(vkey.verify(&body, &sig)?);
+            }
+            ((ECPK, Some(P384)), ES384) => {
+                use p384::ecdsa::signature::Verifier;
+                let vkey = p384::ecdsa::VerifyingKey::from_sec1_bytes(self.subject_public_key)?;
+                let sig = p384::ecdsa::Signature::from_der(sign)?;
+                return Ok(vkey.verify(&body, &sig).unwrap());
+            }
             ((RSA, None), (ID_RSASSA_PSS, Some(p))) => {
+                use signature::{Signature, Verifier};
                 // Decompose the RSA PSS parameters.
                 let RsaSsaPssParams {
-                    hash_algorithm: hash,
+                    hash_algorithm: _hash,
                     mask_algorithm: mask,
-                    salt_length: salt,
-                    trailer_field: tfld,
+                    salt_length: _salt,
+                    trailer_field: _tfld,
                 } = p.decode_into()?;
 
-                // Validate the sanity of the mask algorithm.
-                let algo = match (mask.oid, mask.parameters) {
-                    (ID_MGF_1, Some(p)) => {
-                        let p = p.decode_into::<AlgorithmIdentifier<'_>>()?;
-                        match (p.oids()?, salt, tfld) {
-                            ((SHA256, None), 32, 1) => Ok(SHA256),
-                            ((SHA384, None), 48, 1) => Ok(SHA384),
-                            ((SHA512, None), 64, 1) => Ok(SHA512),
-                            _ => Err(anyhow!("unsupported")),
-                        }
-                    }
-                    _ => Err(anyhow!("unsupported")),
-                }?;
```

Comment:
It is important not to remove this check.

Reply:
Which check is that? The outer match (EC vs RSA) has the default case, the inner match for SHA hashes for RSA has the default case, and the RSA mask.parameters has the `ok_or()` check.

Reply:
We are also checking the size of the `salt` and `trailer`.

Basically, ID_MGF_1 allows signers to specify crazy parameters. We don't. We require only sane combinations.

For example, we support `SHA256` with no parameters, a `salt_length` of `32` and a `trailer_field` of `1`. We reject all other variants of `SHA256`.

Reply:
You should basically keep this match statement as it exists today. It isn't `ring` specific but rather `X.509` specific.

Reply:
It's back